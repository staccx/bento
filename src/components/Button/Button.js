import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  color,
  borderRadius,
  spacing,
  targetSize,
  font,
  fontFamily,
  fontWeight,
  applyVariants,
  commonPropTypes
} from "../../theming"
import themeProps from "./Button.themeProps"
import { componentCreateFactory } from "../../theming/utils/createVariantsFunctionFactory"

const ButtonComp = styled.button`
  background-color: ${color.primary};
  color: ${color.white};
  border-width: 0;
  border-radius: ${borderRadius};
  padding: ${spacing.tiny} ${spacing.medium};
  cursor: pointer;
  min-height: ${targetSize.normal};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  line-height: 32px;
  font-weight: ${fontWeight.bold};
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  transition: background 0.2s ease;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${color.primary};
    color: ${color.white};
  }

  ${applyVariants(themeProps.button)};
`
/**
 * Button component
 */
const Button = props => <ButtonComp {...props} />

const defaultProps = {
  children: "button",
  onClick: null
}

const propTypes = {
  children: commonPropTypes.children.isRequired,
  /**
   * Click callback. Lets you do stuff onClick
   * @export
   */
  onClick: PropTypes.func
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes
Button.themeProps = themeProps
Button.createVariants = componentCreateFactory(Button)

export default Button
