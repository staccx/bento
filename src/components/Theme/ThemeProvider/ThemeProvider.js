import React from "react"
import { ThemeProvider as StyledThemeProvider } from "styled-components"
import PropTypes from "prop-types"
import WebFonts from "../WebFonts"
import GlobalStyle from "../GlobalStyle"
import { theme as instance } from "../../../theming"
import { setLogLevel } from "../../../theming/theme.logger"

/**
 * ThemeProvider is used to wrap you app. It will provide children with correct context
 * It also ensures usage of GlobalStyle and WebFonts
 * @param children
 * @param theme (optional)
 * @return {*}
 */
const ThemeProvider = ({ children, theme, level }) => {
  React.useEffect(() => {
    if (level ?? false) {
      setLogLevel(level)
    }
  }, [level])
  return (
    <StyledThemeProvider theme={theme}>
      <>
        <GlobalStyle />
        <WebFonts />
        {children}
      </>
    </StyledThemeProvider>
  )
}

ThemeProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
    PropTypes.func,
    PropTypes.string
  ]).isRequired,
  /**
   * Theme to override instance
   */
  theme: PropTypes.object,
  level: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
ThemeProvider.defaultProps = {
  theme: instance
}

export default ThemeProvider
