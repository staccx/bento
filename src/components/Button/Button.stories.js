import React from "react"
import Button from "./Button"

export default {
  title: "Components/Button",
  component: Button
}

export const text = () => <Button> Button </Button>
