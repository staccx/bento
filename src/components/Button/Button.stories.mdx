import {useState} from "react"
import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { action }Â from "@storybook/addon-actions"
import Button from "./Button"

<Meta title="components/Button" component={Button} parameters={{component: Button.themeProps.button.name}} />

# Button
Bento Button Component

```javascript
import {Button} from "@staccx/bento"
```


## Standard Button
This is the standard button

<Preview>
  <Story name="Standard">
   <Button>Knapp</Button>
  </Story>
</Preview>


## Long String
<Preview>
  <Story name="Long String">
   <Button>This is a button with long text</Button>
  </Story>
</Preview>

## With useState
<Preview>
  <Story name="with counter">
    {() => {
      const [counter, setCounter] = useState(0);
      const label = `Testing: ${counter}`;
      return <Button onClick={() => setCounter(counter + 1)}>{label}</Button>;
    }}
  </Story>
</Preview>

### Source code

<details><summary>Button</summary>
<p>

```bash
import PropTypes from "prop-types"
import React from "react"
import styled from "styled-components"
import {
  color,
  borderRadius,
  spacing,
  targetSize,
  font,
  fontFamily,
  fontWeight,
  applyVariants,
  commonPropTypes
} from "../../theming"
import themeProps from "./Button.themeProps"
import { componentCreateFactory } from "../../theming/utils/createVariantsFunctionFactory"

const ButtonComp = styled.button`
  background-color: ${color.primary};
  color: ${color.white};
  border-width: 0;
  border-radius: ${borderRadius};
  padding: ${spacing.tiny} ${spacing.medium};
  cursor: pointer;
  min-height: ${targetSize.normal};
  text-decoration: none;
  text-align: center;
  display: inline-block;
  font-size: ${font.base};
  font-family: ${fontFamily.body};
  line-height: 32px;
  font-weight: ${fontWeight.bold};
  white-space: nowrap;
  justify-content: center;
  align-items: center;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  transition: background 0.2s ease;

  &:hover,
  &:focus {
    outline: none;
    background-color: ${color.primary};
    color: ${color.white};
  }

  ${applyVariants(themeProps.button)};
`
/**
 * Button component. Use to click stuff
 */
const Button = props => <ButtonComp {...props} />

const defaultProps = {
  children: "button",
  onClick: null
}

const propTypes = {
  children: commonPropTypes.children.isRequired,
  /**
   * Click callback. Lets you do stuff onClick
   * @export
   */
  onClick: PropTypes.func
}

Button.defaultProps = defaultProps
Button.propTypes = propTypes
Button.themeProps = themeProps
Button.createVariants = componentCreateFactory(Button)

export default Button

```

</p>
</details>


## Props for button
<Props of={Button} />
