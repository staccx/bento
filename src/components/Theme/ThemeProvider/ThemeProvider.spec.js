import renderer from "react-test-renderer"
import React from "react"
import ThemeProvider from "./ThemeProvider"
import Button from "../../Button/Button"
import bentoStorybookTheme from "../../../../.storybook/bentoStorybookTheme"

describe("Theme Provider", () => {
  describe("Button", () => {
    it("Button", () => {
      const tree = renderer
        .create(
          <ThemeProvider>
            <Button>Click me</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("logLevel", () => {
      const tree = renderer
        .create(
          <ThemeProvider level={1}>
            <Button>Click me</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("everything", () => {
      const tree = renderer
        .create(
          <ThemeProvider level={1} theme={bentoStorybookTheme}>
            <Button>Click me</Button>
          </ThemeProvider>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
