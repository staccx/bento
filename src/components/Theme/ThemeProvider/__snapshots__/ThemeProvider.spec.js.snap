// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme Provider Button Button 1`] = `
<button
  className="sc-AxjAm kHAlBU"
  onClick={null}
>
  Click me
</button>
`;

exports[`Theme Provider Button everything 1`] = `
<button
  className="sc-AxjAm kHAlBU"
  onClick={null}
>
  Click me
</button>
`;

exports[`Theme Provider Button logLevel 1`] = `
<button
  className="sc-AxjAm kHAlBU"
  onClick={null}
>
  Click me
</button>
`;
